pkg/catalog/catalog.go

```diff
@@ -1043,90 +1048,103 @@ func (c *Catalog) listCommitsWithPaths(ctx context.Context, repository *graveler
+	outCh := make(chan *commitLogJob, numReadResults)
+	var mgmtGroup multierror.Group
 	mgmtGroup.Go(func() error {
-		num := 0
-		for ; it.Next(); num++ {
-			numm := num
-			vall := it.Value()
-			if done.Load() {
-				break
+		defer close(outCh)
+
+		// workers to check if commit record in path
+		workerGroup, ctx := c.workPool.GroupContext(ctx)
+
+		current := 0
+	readLoop:
+		for it.Next() && !done.Load() {
+			// if context canceled we stop processing
+			select {
+			case <-ctx.Done():
+				break readLoop
+			default:
+			}
+
+			commitRecord := it.Value()
+			// skip merge commits
+			if len(commitRecord.Parents) != NumberOfParentsOfNonMergeCommit {
+				continue
 			}
+
+			// submit work to the pool
+			commitOrder := current
+			current++
 			workerGroup.Submit(func() error {
-				if done.Load() {
-					out <- compareJobResult{num: numm}
-					return nil
+				pathInCommit, err := c.checkPathListInCommit(ctx, repository, commitRecord, paths, commitCache)
+				if err != nil {
+					return err
 				}
-
-				if len(paths) > 0 {
-					// verify that commit includes a change with object/prefix from PathList
-					if len(vall.Parents) != NumberOfParentsOfNonMergeCommit {
-						// skip merge commits
-						out <- compareJobResult{num: numm}
-						return nil
-					}
-					pathInCommit, err := c.checkPathListInCommit(ctx, repository, vall, paths, commitCache)
-					if err != nil {
-						out <- compareJobResult{num: numm, err: err}
-						return err
-					}
-					if !pathInCommit {
-						// no object or prefix found skip commit
-						out <- compareJobResult{num: numm}
-						return nil
-					}
+				job := &commitLogJob{order: commitOrder}
+				if pathInCommit {
+					job.log = convertCommit(commitRecord)
 				}
-
-				out <- compareJobResult{num: numm, log: convertCommit(vall)}
+				outCh <- job
 				return nil
 			})
 		}
-		createdTasks.Store(int64(num))
+		// wait until workers are done and return the first error
+		if err := workerGroup.Wait(); err != nil {
+			return err
+		}
 		return it.Err()
 	})
 
-	results := map[int]*CommitLog{}
+	// process out channel to keep order into results channel by using heap
+	resultCh := make(chan *CommitLog, numReadResults)
+	var jobsHeap commitLogJobHeap
 	mgmtGroup.Go(func() error {
-		for {
-			select {
-			case <-ctx.Done():
-				return nil
-			case res, ok := <-out:
-				if !ok {
-					// compareJobResults is closed, workers finished.
-					return nil
-				}
-				if res.err != nil {
-					drain(done, out)
-					return res.err
-				}
-				results[res.num] = res.log
-
-				// scan the results array to see if we got the number of results needed
-				_, foundAll := findAllCommits(results, params, createdTasks)
-				if foundAll {
-					drain(done, out)
-					return nil
+		defer close(resultCh)
+		// read and sort by heap the result to results channel
+		current := 0
+		for result := range outCh {
+			heap.Push(&jobsHeap, result)
+			for len(jobsHeap) > 0 && jobsHeap[0].order == current {
```

Comment:
You might get stuck forever here, the `results` channel is not ordered. There's no guarantee that `current` will be at the top of the queue.